worlds/stardew_valley/logic/building_logic.py

```diff
@@ -25,78 +23,38 @@ def __init__(self, *args, **kwargs):
+class BuildingLogic(BaseLogic[Union[BuildingLogicMixin, RegionLogicMixin, ReceivedLogicMixin, HasLogicMixin, SourceLogicMixin]]):
 
     @cache_self1
-    def has_building(self, building: str) -> StardewRule:
-        # Shipping bin is special. The mod auto-builds it when received, no need to go to Robin.
-        if building is Building.shipping_bin:
-            if not self.options.building_progression & BuildingProgression.option_progressive:
-                return True_()
-            return self.logic.received(building)
-
-        carpenter_rule = self.logic.building.can_construct_buildings
-        if not self.options.building_progression & BuildingProgression.option_progressive:
-            return Has(building, self.registry.building_rules, has_group) & carpenter_rule
+    def can_build(self, building_name: str) -> StardewRule:
+        building = self.content.farm_buildings.get(building_name)
+        assert building is not None, f"Building {building_name} not found."
 
-        count = 1
-        if building in [Building.coop, Building.barn, Building.shed]:
-            building = f"Progressive {building}"
-        elif building.startswith("Big"):
-            count = 2
-            building = " ".join(["Progressive", *building.split(" ")[1:]])
-        elif building.startswith("Deluxe"):
-            count = 3
-            building = " ".join(["Progressive", *building.split(" ")[1:]])
-        return self.logic.received(building, count) & carpenter_rule
+        source_rule = self.logic.source.has_access_to_any(building.sources)
+        if not building.is_upgrade:
+            return source_rule
 
-    @cached_property
-    def can_construct_buildings(self) -> StardewRule:
-        return self.logic.region.can_reach(Region.carpenter)
+        upgrade_rule = self.logic.building.has_building(building.upgrade_from)
+        return self.logic.and_(upgrade_rule, source_rule)
 
     @cache_self1
-    def has_house(self, upgrade_level: int) -> StardewRule:
-        if upgrade_level < 1:
-            return True_()
+    def has_building(self, building_name: str) -> StardewRule:
+        building_progression = self.content.features.building_progression
 
-        if upgrade_level > 3:
-            return False_()
+        if building_name in building_progression.starting_buildings:
+            return true_
 
-        carpenter_rule = self.logic.building.can_construct_buildings
-        if self.options.building_progression & BuildingProgression.option_progressive:
-            return carpenter_rule & self.logic.received(f"Progressive House", upgrade_level)
+        if not building_progression.is_progressive:
+            return self.logic.building.can_build(building_name)
 
-        if upgrade_level == 1:
-            return carpenter_rule & Has(Building.kitchen, self.registry.building_rules, has_group)
+        # Shipping bin is special. The mod auto-builds it when received, no need to go to Robin.

```

Comment:
Should we make the pet bowl, farm house, basically all the buildings that are "starting buildings" in vanilla make that same assumption? I can make them autobuild from the mod (and, in the case of the first house, I will be forced to, as Robin will have no UI for it)

Reply:
I guess